serial.c

```diff
@@ -452,6 +490,235 @@ static void Interrupt com_general_isr(void)
 /* ======================================================================== */
 
 
+#ifdef PROTECTED_MODE
+
+/* we do not know the exact size of com_general_isr() function
+   but for sure it's not longer then 2048 bytes */
+#define ISR_SIZE        2048
+
+#if defined( __DJGPP__ )
+
+static void serial_dpmi_get_pvect(int vector, _go32_dpmi_seginfo *info)
+{
+    _go32_dpmi_get_protected_mode_interrupt_vector(vector, info);
+}
+
+static void serial_dpmi_set_pvect(int vector, _go32_dpmi_seginfo *info)
+{
+    _go32_dpmi_set_protected_mode_interrupt_vector(vector, info);
+}
+
+static int serial_dpmi_lock_memory(void)
+{
+    unsigned long dataaddr, codeaddr;
+    __dpmi_meminfo dataregion, coderegion;
+
+    if(__dpmi_get_segment_base_address(_my_cs(), &codeaddr) == 0 &&
+       __dpmi_get_segment_base_address(_my_ds(), &dataaddr) == 0)
+    {
+        coderegion.handle = 0;
+        coderegion.size = ISR_SIZE;
+        coderegion.address = codeaddr + (unsigned long)com_general_isr;
+        dataregion.handle = 0;
+        dataregion.size = sizeof(g_comports);
+        dataregion.address = codeaddr + (unsigned long)g_comports;

```

Comment:
On Tue, 22 Dec 2015, Karl Stenerud wrote:

Hi,

> > +    if(__dpmi_get_segment_base_address(_my_cs(), &codeaddr) == 0 &&
> > +       __dpmi_get_segment_base_address(_my_ds(), &dataaddr) == 0)
> > +    {
> > +        coderegion.handle = 0;
> > +        coderegion.size = ISR_SIZE;
> > +        coderegion.address = codeaddr + (unsigned long)com_general_isr;
> > +        dataregion.handle = 0;
> > +        dataregion.size = sizeof(g_comports);
> > +        dataregion.address = codeaddr + (unsigned long)g_comports;
> Can these be cast to `intptr_t` or 'uintptr_t` instead of `unsigned long`?
> Pointer arithmetic via integer types is undefined.

Here we have small problem. In POSIX systems it was documented that long
should be large enough to hold pointers and such casting is commonly used.
It was standard for years.
The only one popular platform which does not confirm this standard and
uses linear addressing is 64bit MS-Windows where [U]LONG is 32bit integer.
It created very serious problems and AFAIR (I'm not very familiar with
DOS/Windows - I usually work with *nixes) to resolve it MS depreciated
commonly used in MS-Windows world casting to [U]LONG and switched to new
type [U]INT_PTR. [u]intptr_t was introduced with C99 as an optional type
in <stdint.h> and older C compiler does not support it. In this particular
case we are talking only about DOS GCC builds (DJGPP). The recent version
of DJGPP supports [u]intptr_t but older ones not just like most of other DOS
compilers (BCC, Watcom, ...). GCC does not even support 'far' modifier for
pointers so it does not have MK_FP(), FP_SEG() and FP_OFF() macros but only
set of inline functions for direct operations on real mode DOS memory like
_farpeekw(). Casting to 'unsigned long' is documented and expected method
of conversion pointers to integers, i.e. this is __dpmi_meminfo definition:
   typedef struct {
      unsigned long handle;
      unsigned long size;
      unsigned long address;
   } __dpmi_meminfo;
so current code casts to expected destination type.
If it looks odd for you then I suggest to rather use macro like:
   #define FP_ADDR(p)         ((unsigned long)(p))
and then:
   coderegion.address = codeaddr + FP_ADDR(com_general_isr);
instead of forcing explicit uintptr_t usage. In DOS builds old C compilers
are still in use and only few modern ones are still developed (DJGPP/GCC,
OpenWatcom, DMC).
There are also other types like ptrdiff_t and size_t. They are more
portable and both will work correctly in this particular case anyhow
their meaning is a little bit different and may confuse users too.

best regards,
Przemek